pkg/printer/printer.go

```diff
@@ -384,6 +384,165 @@ func (p *Printer) WriteProtoNodeStatusEvent(r *observerpb.GetFlowsResponse) erro
 	return nil
 }
 
+func formatServiceAddr(a *pb.ServiceUpsertNotificationAddr) string {
+	return fmt.Sprintf("%s:%d", a.Ip, a.Port)
+}
+
+func getAgentEventDetails(e *pb.AgentEvent) string {
+	switch e.GetType() {
+	case pb.AgentEventType_AGENT_EVENT_UNKNOWN:
+		if u := e.GetUnknown(); u != nil {
+			return fmt.Sprintf("type: %s, notification: %s", u.Type, u.Notification)
+		}
+	case pb.AgentEventType_AGENT_STARTED:
+		if a := e.GetAgentStart(); a != nil {
+			return fmt.Sprintf("start time: %s", fmtTimestamp(a.Time))
+		}
+	case pb.AgentEventType_POLICY_UPDATED, pb.AgentEventType_POLICY_DELETED:
+		if p := e.GetPolicyUpdate(); p != nil {
+			return fmt.Sprintf("labels: [%s], revision: %d, count: %d",
+				strings.Join(p.Labels, ","), p.Revision, p.RuleCount)
+		}
+	case pb.AgentEventType_ENDPOINT_REGENERATE_SUCCESS, pb.AgentEventType_ENDPOINT_REGENERATE_FAILURE:
+		if r := e.GetEndpointRegenerate(); r != nil {
+			str := fmt.Sprintf("id: %d, labels: [%s]", r.Id, strings.Join(r.Labels, ","))
+			if re := r.Error; re != "" {
+				str += fmt.Sprintf(", error: %s", re)
+			}
+			return str
+		}
+	case pb.AgentEventType_ENDPOINT_CREATED, pb.AgentEventType_ENDPOINT_DELETED:
+		if ep := e.GetEndpointUpdate(); ep != nil {
+			str := fmt.Sprintf("id: %d", ep.Id)
+			if n := ep.PodName; n != "" {
+				str += fmt.Sprintf(", pod name: %s", n)
+			}
+			if n := ep.Namespace; n != "" {
+				str += fmt.Sprintf(", namespace: %s", n)
+			}
+			return str
+		}
+	case pb.AgentEventType_IPCACHE_UPSERTED, pb.AgentEventType_IPCACHE_DELETED:
+		if i := e.GetIpcacheUpdate(); i != nil {
+			str := fmt.Sprintf("cidr: %s, identity: %d", i.Cidr, i.Identity)
+			if i.OldIdentity != nil {
+				str += fmt.Sprintf(", old identity: %d", i.OldIdentity.Value)
+			}
+			if i.HostIp != "" {
+				str += fmt.Sprintf(", host ip: %s", i.HostIp)
+			}
+			if i.OldHostIp != "" {
+				str += fmt.Sprintf(", old host ip: %s", i.OldHostIp)
+			}
+			return str + fmt.Sprintf(", encrypt key: %d", i.EncryptKey)
+		}
+	case pb.AgentEventType_SERVICE_UPSERTED:
+		if svc := e.GetServiceUpsert(); svc != nil {
+			str := fmt.Sprintf("id: %d", svc.Id)
+			if fe := svc.FrontendAddress; fe != nil {
+				str += fmt.Sprintf(", frontend: %s", formatServiceAddr(fe))
+			}
+			if bes := svc.BackendAddresses; len(bes) != 0 {
+				backends := make([]string, 0, len(bes))
+				for _, a := range bes {
+					backends = append(backends, formatServiceAddr(a))
+				}
+				str += fmt.Sprintf(", backends: [%s]", strings.Join(backends, ","))
+			}
+			if t := svc.Type; t != "" {
+				str += fmt.Sprintf(", type: %s", t)
+			}
+			if tp := svc.TrafficPolicy; tp != "" {
+				str += fmt.Sprintf(", traffic policy: %s", tp)
+			}
+			if n := svc.Name; n != "" {
+				str += fmt.Sprintf(", name: %s", n)
+			}
+			if ns := svc.Namespace; ns != "" {
+				str += fmt.Sprintf(", namespace: %s", ns)
+			}
```

Comment:
for performance reasons, instead of joining strings multiple times, a string builder or `bytes.Buffer` should be used

Reply:
Done, thanks.